### ISSUE TITLE
Request for sampling from posterior

### ISSUE THREAD

**Issue by blindedauthor** [GPflow/GPflow#780]
Thank you for your work on this excellent project 

I have a feature request: 

Using the predict_f_samples() method of a GPflow model instance I can sample from the GP latent function but I can't see a way to sample from the posterior, which takes into account the likelihood. 
Will it be possible to add this feature to GPflow?
GPy allows sampling from the posterior, taking into account the likelihood, using the posterior_samples() method of their model.

An example for a Gaussian likelihood
```python
import numpy as np
import matplotlib.pyplot as plt 
import GPy as GPy
import gpflow
np.random.seed(39) # set random seed 
N_SIZE=100
MEAN_SET=10.15
STD_SET=2.33
# Create some Gaussian data
X = np.linspace(0,23.50,N_SIZE)[:, None]
Xpredict=np.linspace(24,30,N_SIZE)[:, None]
Y = np.random.normal(loc=MEAN_SET,scale=STD_SET,size=N_SIZE) [:, None]
# plot data
# Create and fit GPy model
kernelGPy = GPy.kern.Bias(1)
modelGPy=GPy.models.GPRegression(X,Y,kernel=kernelGPy)
modelGPy.optimize()
sampleGPy=modelGPy.posterior_samples(Xpredict,10)
# Create and fit Gflowy model
kernelGPflow = gpflow.kernels.Bias(1)
modelGPflow=gpflow.models.GPR(X,Y,kern=kernelGPflow)
gpflow.train.ScipyOptimizer().minimize(modelGPflow)
sampleGPflowLatent=modelGPflow.predict_f_samples(Xpredict,10).squeeze().T
# plot data
plt.figure()
plt.plot(X,Y,color='tab:red',marker='o',
      linestyle = 'None',label='Raw Data') 
for column in sampleGPy.T:
    plt.plot(Xpredict,column,color='tab:orange',marker='x',
          linestyle = 'None') 
plt.plot(Xpredict,column,color='tab:orange',marker='x',
  linestyle = 'None', label='GPy samples') 
for column in sampleGPflowLatent.T:
    plt.plot(Xpredict,column,color='tab:blue',marker='x',
          linestyle = 'None')
plt.plot(Xpredict,column,color='tab:blue',marker='x',
      linestyle = 'None', label='GPflow samples')   
plt.legend()         
```
![image](https://user-images.githubusercontent.com/32488840/41045901-443460aa-69a1-11e8-9b85-21ac9482d8d1.png)

An example for a Poisson likelihood
```python
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt 
import GPy 
import gpflow
plt.ion()
NUM_SAMPLES=5
# Create some Poisson data
X = np.linspace(0,23.50,48)[:, None]
intensities = 4 + 3*np.sin(10 + 0.6*X) + np.sin(0.1*X)
Y = np.array([sp.random.poisson(intensity) for intensity in intensities])
# Fit GPy Model
modelGPy = GPy.core.GP(X=X, Y=Y, likelihood=GPy.likelihoods.Poisson(), 
                     inference_method=GPy.inference.latent_function_inference.Laplace(), 
                     kernel=GPy.kern.RBF(1))
modelGPy.optimize()
# Fit GPflow Model
modelGPFlow =  gpflow.models.VGP(X=X,Y=Y.astype('float64'),kern=gpflow.kernels.RBF(1),
                              likelihood=gpflow.likelihoods.Poisson())
gpflow.train.ScipyOptimizer().minimize(modelGPFlow)
# Samples from GPflow model
sampleGPflow=modelGPFlow.predict_f_samples(X,NUM_SAMPLES).squeeze().T
# Samples from GPy model
sampleGPy=modelGPy.posterior_samples(X,NUM_SAMPLES)

# Plot samples and original data
plt.figure()
for column in sampleGPflow.T:
    plt.plot(X,column,color='tab:blue',marker='x',
      linestyle = 'None') 
plt.plot(X,column,color='tab:blue',marker='x',
  linestyle = 'None',label='GPflow')     
for column in sampleGPy.T:
    plt.plot(X,column,color='tab:orange',marker='x',
      linestyle = 'None')  
plt.plot(X,column,color='tab:orange',marker='x',
  linestyle = 'None',label='GPy')     
plt.scatter(X,Y,color='tab:red',label='Data')
plt.legend()
```
![image](https://user-images.githubusercontent.com/32488840/40983284-2baeebb0-68d7-11e8-97f8-649437e50f0b.png)

GPflow version 1.1.1
tensorflow version 1.8.0
no GPU used
Python 3.6.5
Windows 10

**Comment by hughsalimbeni** [comment]
Doing this on the tensorflow side would be tricky, though actually to my surprise I see that tensorflow now has `random_poission` and `random_gamma` which covers the two hardest ones. We would need to implement multiclass, however, as  `tf.multinomial` is softmax, and beta is not implemented. The question is how necessary is this feature, as it can always be done easily on the python side. @markvdw  any thoughts?

**Comment by Joshuaalbert** [comment]
Just in case, there is multiclass sampling in TF [here](https://www.tensorflow.org/api_docs/python/tf/distributions/Categorical).

**Comment by blindedauthor** [comment]
Thanks for the replies. @hughsalimbeni you are right, I am currently doing the transformation of the samples for Gaussian and Poisson likelihoods outside of GPflow using Numpy. I think it will be a useful feature. For example in anomaly detection you might want to compare the samples drawn from the GP to some hold-out data to see if there is an anomaly.
